### ISSUE TITLE
Simplify or eliminate the C stubs needed for supplemental modules

### ISSUE THREAD

**Issue by PaulWessel** [GenericMappingTools/gmt#3130]
**Description of the desired feature**

For plugins used in GMT, one must prepare files similar to gmt_core_modules.[ch], but for the custom supplement.  In GMT, we create this via the command `gmt_make_module_src.sh core`, which builds the necessary source files - these are also maintain in the repo.  This is not very new developer-friendly, prone to scripting mistakes (that script is pushing 500 lines), and seems very repetitive for different shared libraries.  I will try to place as much of these items in an abstract way into the regular gmt library and hope that a developer only has to specify a few things in their CMakeLists.txt file.  I will open a branch to explore this goal.  I would things to work this way:

Instead of separate function called gmtlib_core_module_list_all, gmtlib_supplements_module_list_all, gmtlib_gsfml_module_list_all, etc. there should only be a single function in the GMT library called gmtlib_module_list_all and it should accept the Gmt_moduleinfo struct pointer from each shared library.  Supplement maintainers would then only be responsible for a small include file with specific information for each of their modules.

**Comment by PaulWessel** [comment]
This is close to being completed via #3133.  A few questions remain:

Truly external shared libraries, such as GSFML and MB-System, may have very different build systems.  However, if the supplements follow a few simple design rules they should all be able to tap into the GMT CMake build (?).

**Design Rules**

1. Each module.c file must set these preprocessor parameters (here example from psclip.c):

```
#define THIS_MODULE_CLASSIC_NAME    "psclip"
#define THIS_MODULE_MODERN_NAME     "clip"
#define THIS_MODULE_LIB             "core"
#define THIS_MODULE_PURPOSE         "Initialize or terminate polygonal clip paths"
#define THIS_MODULE_KEYS            "<D{,>X},C-("
```

Depending on integration with GMT, the KEYS can be set to blank "".  For most external supplements the CLASSIC and MODERN names will be identical.  The LIB names the external shared library group.

2. Each shared library needs to configure two files:

- gmt_suppl_moduleinfo.h: The structure contents generated from the above preprocessor parameters.
- gmt_suppl_glue.c: Configured from GMT's gmt_glue.c.in with SUPPL_NAME set to the name of the supplement (here suppl).

We expect these two files will easily be generated by the supplements CMakeLists.txt file if CMake is used.  However, what about other supplements (e.g., mbsystem) that uses autoconf? Do we need to supply a helper script (again) via gmt/share/tools that generates these two files? These files need to be included in the mb build process that results in mbsystem.so.

Currently, these two files are included with the GMT core.  This is not good since the GMT core should not have any knowledge about MB-system specific.

**Comment by PaulWessel** [comment]
The CMake solution we use for the official supplements and others that are configured adn built with gmt may not work well with other supplements (e.g., mbsystem) that expects GMT to be built already.  A cross-platform, cross-build-system solution might be to let a new gmt --build-glue=suppl command create the two files by examining the C files in the current directory.  Then,mbsystem makefile simply needs to run gmt --build-glue=mbsystem before building the shared library.  From the third party users point of view this must surely be simple (just run a command during configuration),  compared to us supplying an autoconf macro, a CMake macro, and possibly a shell script?

**Comment by PaulWessel** [comment]
FYI @joa-quim and @seisman: I have summarized some notes from the now closed PR.  There are three kinds of "supplements":

1. Official ones in GMT, like _seis_, _spotter_... These are built at the **same time** as GMT.
2. Unofficial ones, like _gsfml_, that take advantage of the GMT build machinery by setting a symbolic link newsuppl[?] to point to that supplements src dir. They are then also built at the **same time** as GMT.
3. Unofficial ones, like _mbsystem_, that has its own build system that expects GMT to be installed already.

For type 2 above to work, the idea was to use configure_file with gmt_glue.c.in and make src/newsuppl/gmt_gsfml_glue.c which means there has to be a gmt_gsfml_moduleinfo.h available. It would be slick if that was handled by GMT CMake, but if it adds to much complication... Then the 3rd option, and possibly the solution for type 3, is to let `gmt --build-glue=supplementname` create those pieces.

Note that there is **no requirement** that an external GMT supplement be built with the glue.  However, there are two good reasons to enable it:

1. Visibility via gmt --help or gmt --show-modules. Without the glue functions we cannot discover modules in a shared library. You would need to know the module names and look up syntax online.
2. Building external interfaces. If GMT/MATLAB toolbox (or Julia I think) wanted to call some of these foreign modules and take advantage of matlab syntax, e.g.,

```
G = gmt ('surface', 'R0/30/0/30 -I0.1 mbdata.txt');  %grid a large MB dataset
plot = gmt ('mbgrdtiff', 'some options', G);  % Make tiff file via mbsystem
```

then we need to be able to access the KEYS for those modules in the library. Otherwise we cannot do the above since we would not know how to hook the G matrix into the mbgrdtiff command line options. This is the job of API functions _GMT_Encode_Options_ and _GMT_Expand_Option_.

So that is why I want to encourage having the glue **and** making it super-painless for those developers to do this.

**Comment by PaulWessel** [comment]
So I have C code ready to test that would allow a `gmt --new-glue=suppl` command to create a single gmt_suppl_glue.c with the inline content of the keys structure.  I don't think we can make life much more simpler for external developers using autoconf or plain makefiles, no? Writing automake macros and having the developers include them is likely to meet more resistance than saying "just run this command in your makefile".  After all, these developers like mbsystem) does not really need GMT to gain access to a shared library so their incentive is low.  Shall I give i a spin, @seisman ?

**Comment by seisman** [comment]
Currently, the information of type 2 modules (e.g. gsfml) are listed in gmt_supplements_moduleinfo.h, but the modules are built into a separate library (e.g. gsfml.so). Can you access your gsfml modules in GMT/Matlab?

For type 3, I agree `gmt --new-glue=suppl` may be the simplest way.

**Comment by PaulWessel** [comment]
I am trying to build gsfml again via type 2 after all our changes.  I have a file that is configured in src/newsuppl/CMakeLists.txt:

configure_file (gsfml_config.h.in gsfml_config.h)

I have having trouble listing that include file in SUPPL_PROGS_SRCS.  I tried

`	${CMAKE_CURRENT_BINARY_DIR}/newsuppl/gsfml_config.h)`

but get

```
CMake Error at src/CMakeLists.txt:750 (add_library):
  Cannot find source file:

    newsuppl/Users/pwessel/GMTdev/gmt-dev/dbuild/src/newsuppl/newsuppl/gsfml_config.h
```

Apart from fixing this, it is of course not ideal to have to type newsuppl into the gsfml Cmake files.

**Comment by PaulWessel** [comment]
I should say the configured file is created and is in build/src/newsuppl/gsfml_config.h so that part works.

**Comment by PaulWessel** [comment]
Taking out the newssuppl link gets closer but it still has that leading newsuppl:

    newsuppl/Users/pwessel/GMTdev/gmt-dev/dbuild/src/newsuppl/gsfml_config.h

**Comment by seisman** [comment]
What's your SUPPL_PROGS_SRCS setting?

**Comment by PaulWessel** [comment]
```
set (SUPPL_PROGS_SRCS fz_analysis.h fzanalyzer.c
	fzblender.c mlconverter.c Chron_Normal.h Chron_Reverse.h
	Geek2007n.h Geek2007r.h CK1995n.h CK1995r.h GST2004n.h GST2004r.h
	${CMAKE_CURRENT_BINARY_DIR}/gsfml_config.h)

```

**Comment by seisman** [comment]
OK, I see the problem. We use 
```
get_subdir_var_files (_suppl_lib_srcs SUPPL_LIB_SRCS ${_dir})
```
to get SUPPL_PROGS_SRCS but prepend the directory name before the files. So the variable **_suppl_lib_srcs** is a list of `newsuppl/fz_analysis.h`, `newsuppl/fzanalyzer.c`. Of course, it doesn't work for absolute paths like ${CMAKE_CURRENT_BINARY_DIR}/gsfml_config.h.

**Comment by seisman** [comment]
I think it's a bug of the `get_subdir_var_files` macro. We should check if a file is an absolute path before prepending a directory name.

**Comment by PaulWessel** [comment]
The whole idea with the symbolic link newsuppl[1-9] was to avoid having to place things in gmt-dev/src.  So .gitignore lists newsuppl?.  Perhaps this is not a good solution.
What if in the Advanced.cmake we set 

`set (SUPPL_EXTRA_DIRS /Users/pwessel/GMTdev/gsfml/TOOLS/src)`

instead.  That makes more sense anyway; then we can toss the newsuppl junk in .gitingore.

**Comment by seisman** [comment]
> `set (SUPPL_EXTRA_DIRS /Users/pwessel/GMTdev/gsfml/TOOLS/src)`

It sounds a good idea, but I'm afraid you will see the same issue like:
``` 
/Users/pwessel/GMTdev/gsfml/TOOLS/src//Users/pwessel/GMTdev/gmt-dev/dbuild/src/newsuppl/gsfml_config.h not found
```

**Comment by PaulWessel** [comment]
Yes true, the question is if there is a solution to that problem.

**Comment by PaulWessel** [comment]
Hi @seisman and @joa-quim. Where should I write up the various options and solutions for external developers who want to include their modules as GMT supplements, i.e., types 1-4?  Maybe a new SUPPLEMENT_BUILDING.md file in the top level?  I think the API docs is too developer-specific, but it could also reference it.

**Comment by joa-quim** [comment]
API docs seems the most logical place.

**Comment by seisman** [comment]
I prefer a separate reST document in the doc/rst directory and list it under the "Developer Resources". 
![image](https://user-images.githubusercontent.com/3974108/80321716-4c942e80-87ed-11ea-8268-c790abce4385.png)

**Comment by PaulWessel** [comment]
Hi @seisman, the documentation part of this issue is separate from the implementation of the C stubs, and is now being addressed in #3317 anyway.  Can we close this one?
